// async-genetic/lib/genetic.esm.js
function clone(obj) {
    if (obj == null || typeof obj != 'object')
        return obj;
    return JSON.parse(JSON.stringify(obj));
}
function Maximize(a, b) {
    return a >= b;
}
function Minimize(a, b) {
    return a < b;
}

const Select = { Tournament2, Tournament3, Fittest, Random, RandomLinearRank, Sequential };
class Genetic {
    constructor(options, entities = []) {
        this.entities = entities;
        this.stats = {};
        this.internalGenState = {}; /* Used for random linear */
        this.population = [];
        this.tryCrossover = () => {
            const { crossoverProbablity, crossoverFunction } = this.options;
            let selected = crossoverFunction && Math.random() <= crossoverProbablity ? this.selectPair() : this.selectOne();
            if (selected.length === 2) {
                selected = crossoverFunction(selected[0], selected[1]);
            }
            return selected.map(this.tryMutate);
        };
        this.tryMutate = (entity) => {
            // applies mutation based on mutation probability
            if (this.options.mutationFunction && Math.random() <= this.options.mutateProbablity) {
                return this.options.mutationFunction(entity);
            }
            return entity;
        };
        const defaultOptions = {
            populationSize: 250,
            mutateProbablity: 0.2,
            crossoverProbablity: 0.9,
            fittestNSurvives: 1,
            select1: Select.Tournament2,
            select2: Select.Tournament2,
            optimize: Optimize.Maximize,
        };
        this.options = { ...defaultOptions, ...options };
    }
    seed(entities = []) {
        this.entities = entities;
        // seed the population
        for (let i = 0; i < this.options.populationSize; ++i) {
            this.entities.push(this.options.randomFunction());
        }
    }
    best(count = 1) {
        return this.population.slice(0, count).map((ph) => ph.entity);
    }
    breed() {
        // crossover and mutate
        const newPop = [];
        // lets the best solution fall through
        if (this.options.fittestNSurvives) {
            newPop.push(...this.population.slice(0, this.options.fittestNSurvives).map((ph) => ph.entity));
        }
        // Lenght may be change dynamically, because fittest and some pairs from crossover
        while (newPop.length < this.options.populationSize) {
            newPop.push(...this.tryCrossover());
        }
        this.entities = newPop;
    }
    async estimate() {
        const { fitnessFunction, optimize } = this.options;
        // reset for each generation
        this.internalGenState = {};
        // cleanup score and sort
        this.population.length = 0;
        const tasks = this.entities.map((entity) => fitnessFunction(entity));
        this.population = await (await Promise.all(tasks))
            .map((fitness, i) => ({ fitness, entity: this.entities[i] }))
            .sort((a, b) => (optimize(a.fitness, b.fitness) ? -1 : 1));
        const popLen = this.population.length;
        const mean = this.getMean();
        this.stats = {
            maximum: this.population[0].fitness,
            minimum: this.population[popLen - 1].fitness,
            mean,
            stdev: this.getStdev(mean),
        };
    }
    /**
     * Mean deviation
     */
    getMean() {
        return this.population.reduce((a, b) => a + b.fitness, 0) / this.population.length;
    }
    /**
     * Standart deviation
     */
    getStdev(mean) {
        const { population: pop } = this;
        const l = pop.length;
        return Math.sqrt(pop.map(({ fitness }) => (fitness - mean) * (fitness - mean)).reduce((a, b) => a + b, 0) / l);
    }
    selectOne() {
        const { select1 } = this.options;
        return [clone(select1.call(this, this.population))];
    }
    selectPair() {
        const { select2 } = this.options;
        return [clone(select2.call(this, this.population)), clone(select2.call(this, this.population))];
    }
}
/** Utility */
function Tournament2(pop) {
    const n = pop.length;
    const a = pop[Math.floor(Math.random() * n)];
    const b = pop[Math.floor(Math.random() * n)];
    return this.options.optimize(a.fitness, b.fitness) ? a.entity : b.entity;
}
function Tournament3(pop) {
    const n = pop.length;
    const a = pop[Math.floor(Math.random() * n)];
    const b = pop[Math.floor(Math.random() * n)];
    const c = pop[Math.floor(Math.random() * n)];
    let best = this.options.optimize(a.fitness, b.fitness) ? a : b;
    best = this.options.optimize(best.fitness, c.fitness) ? best : c;
    return best.entity;
}
function Fittest(pop) {
    return pop[0].entity;
}
function Random(pop) {
    return pop[Math.floor(Math.random() * pop.length)].entity;
}
function RandomLinearRank(pop) {
    this.internalGenState['rlr'] = this.internalGenState['rlr'] || 0;
    return pop[Math.floor(Math.random() * Math.min(pop.length, this.internalGenState['rlr']++))].entity;
}
function Sequential(pop) {
    this.internalGenState['seq'] = this.internalGenState['seq'] || 0;
    return pop[this.internalGenState['seq']++ % pop.length].entity;
}
const Optimize = { Minimize, Maximize };

export { Genetic, Select };
